serverless-heaven/serverless-webpack
# Fixed uncontrolled second compile when entering watch mode for offline.

## What did you implement:

Closes #279 



## How did you implement it:



`webpack.compiler.watch()` automatically triggers a compile. This led to a double compile (with the second one being uncontrolled and asynchronous) on startup of the plugin together with serverless offline. The second unwanted compile was running in parallel with the serverless-offline start.

Now, for watch mode, the inital compile is skipped as the watch will do it anyway. For non-watch mode the regular `webpack:compile` lifecycle is triggered.

Additionally a new switch `--webpack-no-watch` has been added to turn off the default watch mode with serverless-offline. This is needed, if you use serverless-offline in an automated build that does integration tests (and changes the source tree) with an external script given to offline. There you only want to build once and test the output with `serverless offline --exec` automatically. As soon as the given script changes the source tree, it leads to a recompile which breaks the script.

## How can we verify it:



Use `serverless offline` or `serverless offline start` and check that there is only one compile in the beginning (and not two) and that the plugin waits for its completion. Only then the offline plugin should start its work (watch mode enabled as before).

Use `serverless offline --webpack-no-watch` to start with watch mode disabled. Changes in the sources should not trigger a rebuild now.

## Todos:


***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO
